### ISSUE TITLE
use some internal go api to read/write log msg

### ISSUE THREAD

**Issue by vtolstov** [logrange/logrange#97]
I want to provide some external api for logrange server that can be used with my own app. How can i work with logrange server without using tcp/ip? So my external api server listens for requests and work with logrange via internal api.
I can use logrange rcp via tcp/ip because all of y aps uses mtls certs and auth via oauth2 compatible server with token to obtain mtls certs.

**Comment by dspasibenko** [comment]
@vtolstov I presume the question relates to embedding logrange to your application. In theory it is quite possible, but some changes are required. You can take a look at [the code](https://github.com/logrange/logrange/blob/1a4caa46d2cebcc288e871cce43ce717ec9c34be/server/server.go#L37) and embed logrange into your app. I am just curious, why do you need it at all? 

you can use TLS with logrange or modify the code for your own auth if you need.

Also, may be you don't need all metadata features, in this case please take a look at [range](https://github.com/logrange/range) which has some programming interfaces to embed journaling functionality into your app.

**Comment by vtolstov** [comment]
i'm use go-micro framework and have 25 microservices. It uses gossip to obtain endpoint (ip:port) of other services.
go-micro able to abstract transport (grpc, tcp, udp, quic...) so i want to have my own service that uses go-micro and registers in registry able to recv logs and put it to logrange. But if i'm strart logrange with listening i need to specify host:port to listen and some tls auth. But my microservices obtain random free ports and registers in registry, so other services knows about new endpoint.
So i'm prefer to be able to use logrange via some go api. Also this is have benefit to you - you can easy create grpc server that listens for reqs and use internal go api to handle it.

**Comment by vtolstov** [comment]
P.S. I'm need indexing/labels feature so as i see pure range package is not that i need.

**Comment by vtolstov** [comment]
@dspasibenko how hard to provide api for read/write logs without rpc overhead?

**Comment by dspasibenko** [comment]
@vtolstov 
>> how hard to provide api for read/write logs without rpc overhead?
If you want to embed it, everything is already there. Just take a look how API translated different calls [here](https://github.com/logrange/logrange/blob/aab41da22db99dce9bff11f001996a3638609480/api/rpc/ingestor.go#L107) or [here](https://github.com/logrange/logrange/tree/master/pkg/backend)

**Comment by dspasibenko** [comment]
@vtolstov please also take into account that if you will use any other transport, but not "vanilla  logrange rpc", most probably you will loose performance and add overhead for sure. Especially if you are going to use HTTP. My guess if you will use GRPC you will loose performance twice, but if it is not critical, not a big deal then.

**Comment by vtolstov** [comment]
thanks! that's fine! Did you mind to provide it via Server interface?

**Comment by vtolstov** [comment]
I don't think that volume of my logs is huge, but i need firstly single control of who can access log server and how. All of my services uses rbac/abac style permissions and as i say - mTLS. I think that bring this to logrange is painful ....
